File: ocaml/2015/day04b/day04b.ml
open Printf
open Sys

let usage progname =
  eprintf "usage: %s <key>\n" progname;
  exit 1

let process_key key =
  let rec check_key key n =
    let try_key = key ^ (string_of_int n) in
    let digest = Digest.MD5.string try_key in
    let hex_bytes = Digest.MD5.to_hex digest in
    if String.starts_with ~prefix:"000000" hex_bytes then
      n
    else
      check_key key (n + 1)
  in check_key key 1

let () =
  let args = Sys.argv
  in if Array.length args < 2 then
       usage args.(0) ;
     let key = args.(1) in
     let result = process_key key in
     printf "result = %d\n" result

File: julia/2015/day04b/day04b.jl
#!/usr/bin/env julia

using MD5

function usage()
    progname = PROGRAM_FILE
    println(stderr, "usage: $progname <key>")
    exit(1)
end

function process(key::String)::Int32
    n = 1
    while true
        try_key = key * string(n)
        digest = md5(try_key)
        hex_digest = bytes2hex(digest)
        if startswith(hex_digest, "000000")
            return Int32(n)
        else
            n += 1
        end
    end
end

function main()
    args = ARGS
    if length(args) < 1
        usage()
    end
    key = args[1]
    result = process(key)
    println("result = $result")
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
